output/sarif.go

```diff
@@ -0,0 +1,496 @@
+	}
+	return processResults(run, []Result{result}, skippedResultType, fileName, namespace, ruleMap)
+}
+
+// Output outputs the results in SARIF format.
+func (s *SARIF) Output(results []CheckResult) error {
+	startTime := time.Now().UTC()
+
+	// Create SARIF report structure
+	report := sarifReport{
+		Schema:  sarifSchemaURI,
+		Version: sarifVersion,
+		Runs: []sarifRun{
+			{
+				Tool: sarifTool{
+					Driver: sarifDriver{
+						Name:           toolName,
+						InformationURI: toolURI,
+						Rules:          []sarifRule{},
+					},
+				},
+				Results:     []sarifResult{},
+				Invocations: []sarifInvocation{},
+			},
+		},
+	}
+
+	run := &report.Runs[0]
+	ruleMap := make(map[string]bool)
+
+	// Process all results
+	for _, result := range results {
+		err := processResults(run, result.Failures, failureResultType, result.FileName, result.Namespace, ruleMap)
+		if err != nil {
+			return fmt.Errorf("process failures: %w", err)
+		}
+		err = processResults(run, result.Warnings, warningResultType, result.FileName, result.Namespace, ruleMap)
+		if err != nil {
+			return fmt.Errorf("process warnings: %w", err)
+		}
+		err = processResults(run, result.Exceptions, exceptionResultType, result.FileName, result.Namespace, ruleMap)
+		if err != nil {
+			return fmt.Errorf("process exceptions: %w", err)
+		}
+
+		// Add success result if no failures/warnings/exceptions
+		if len(result.Failures) == 0 && len(result.Warnings) == 0 && len(result.Exceptions) == 0 {
+			if result.Successes > 0 {
+				err = createSuccessResult(run, result.FileName, result.Namespace, ruleMap)
+				if err != nil {
+					return fmt.Errorf("create success result: %w", err)
+				}
+			} else {
+				err = createSkippedResult(run, result.FileName, result.Namespace, ruleMap)
+				if err != nil {
+					return fmt.Errorf("create skipped result: %w", err)
+				}
+			}
+		}
+	}
+
+	// Add invocation information
+	exitCode := 0
+	exitDesc := exitNoViolations
+	if hasFailures(results) {
+		exitCode = 1
+		exitDesc = exitViolations
+	} else if hasWarnings(results) {
+		exitCode = 0
+		exitDesc = exitWarnings
+	} else if hasExceptions(results) {
+		exitCode = 0
+		exitDesc = exitExceptions
+	}
+
+	run.Invocations = []sarifInvocation{
+		{
+			ExecutionSuccessful: true,
+			ExitCode:            exitCode,
+			ExitCodeDescription: exitDesc,
+			StartTimeUtc:        startTime.Format(time.RFC3339),
+			EndTimeUtc:          time.Now().UTC().Format(time.RFC3339),
+		},
+	}
+
+	// Marshal to JSON
+	encoder := json.NewEncoder(s.writer)
+	encoder.SetIndent("", "  ")
+	if err := encoder.Encode(report); err != nil {
+		return fmt.Errorf("encode sarif: %w", err)
+	}
+
+	return nil
+}
+
+// Report is not supported in SARIF output
+func (s *SARIF) Report(_ []*tester.Result, _ string) error {
+	return fmt.Errorf("report is not supported in SARIF output")
+}
+
+// hasFailures returns true if any of the results contain failures
+func hasFailures(results []CheckResult) bool {
```

Comment:
nit: Go stdlib now has `slices.ContainsFunc` which could be used to reduce repetition in these functions.